### PR TITLE
Ref<T> type and convenience syntax around it

### PR DESCRIPTION
**Pull request by ryzhyk**
- `Ref<T>` is the DDlog shortcut to Rust's `Arc<T>` type
- syntax for declaring relations whose records are references:
 ```
 relation &Router(
     lr:                 nb.Logical_Router,
     dpname:             string,
     l3dgw_port:         Option<nb.Logical_Router_Port>,
     redirect_port_name: string,
     is_gateway:         bool,
     snat_external_ips:  Set<string>
 )
 ```
- Pattern matching relations and structs wrapped in a `Ref`
```
&RouterPort(.lrp = lrp,
                 .router = &Router{.lr = lr, .dpname = dpname, .snat_external_ips = external_ips},
                 .json_name = json_name,
                 .networks = networks)
```
- syntax in flux; no documentation yet

### CODE REVIEW

File: src/Language/DifferentialDatalog/Validate.hs [hunk 1]
```diff
@@ -414,8 +411,13 @@ exprValidate1 _ _ ctx (EPHolder p)        = do
                    CtxStruct EStruct{..} _ f -> "Missing field " ++ f ++ " in constructor " ++ exprConstructor
                    _               -> "_ is not allowed in this context"
     check (ctxPHolderAllowed ctx) p msg
-exprValidate1 d _ ctx (EBinding p v _)    = checkNoVar p d ctx v
+exprValidate1 d _ ctx (EBinding p v _)    = do
+    checkNoVar p d ctx v
+    
 exprValidate1 d tvs _ (ETyped _ _ t)      = typeValidate d tvs t
+exprValidate1 _ _ ctx (ERef p _)          =
+    -- Rust does not allow pattern matching inside 'Arc'
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Validate.hs]: this is a strange comment. what should matter is what ddlog allows, not what rust allows.

File: test/datalog_tests/simple.dl [hunk 2]
```diff
@@ -681,3 +681,24 @@ Allocation(name, id, x) :-
...
+input relation Referenced(x: bool, y: Option<string>)
+
+output relation Referee(r: Ref<Referenced>)
+
+Referee(ref_new(r)) :-
+    r in Referenced(_, Some{_}).
+
+output relation Filtered(r: Ref<Referenced>)
+
+Filtered(r.r) :- r in Referee(.r = &Referenced{.x = true}).
```
Comment (mbudiu-vmw) [inline test/datalog_tests/simple.dl]: Is equality for references well-defined? I suspect they behave like pointers. But this means you cannot just copy them.

Comment (mbudiu-vmw) [reply test/datalog_tests/simple.dl]: I.e., if you delete a value and insert it back again you will have the same value in the relation, but the other relation will have a different reference. I will have to think about the semantics of this.

File: src/Language/DifferentialDatalog/Validate.hs (hunk 1 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Validate.hs]: Sure, but it is really hard to implement things that don't directly map to Rust. The comment just documents the reason for the limitation.

File: test/datalog_tests/simple.dl (hunk 2 above)
Comment (ryzhyk) [reply test/datalog_tests/simple.dl]: The current semantics compares values inside references. In the scenario you described, there should still only be one reference, but there theoretically exist more complex dataflows where this is indeed possible. In any case, multiple references should not affect the functionality; only memory footprint.

**Pull request merged**